src/main/java/com/netflix/imflibrary/writerTools/IMPAssembler.java

```diff
@@ -0,0 +1,387 @@
+                logger.info("track: {}, file: {}: path: {}", simpleTimeline.getTracks().indexOf(track), track.getMediaFiles().indexOf(mediaFile), mediaFile.getFile().getAbsolutePath());
+                ResourceByteRangeProvider resourceByteRangeProvider = new FileByteRangeProvider(mediaFile.getFile());
+                PayloadRecord headerPartitionPayloadRecord = IMPFixer.getHeaderPartitionPayloadRecord(resourceByteRangeProvider, imfErrors);
+                if (headerPartitionPayloadRecord == null) {
+                    throw new IOException("Could not get header partition for file: " + mediaFile.getFile().getAbsolutePath());
+                }
+                byte[] headerPartitionBytes = headerPartitionPayloadRecord.getPayload();
+                byte[] hash = IMFUtils.generateSHA1Hash(resourceByteRangeProvider);
+
+                UUID trackFileId = IMPFixer.getTrackFileId(headerPartitionPayloadRecord);
+                logger.info("UUID read from file: {}: {}", mediaFile.getFile().getName(), trackFileId.toString());
+                logger.info("Adding file {} to imfTrackFileMetadataMap", mediaFile.getFile().getName());
+                imfTrackFileMetadataMap.put(
+                        trackFileId,
+                        new IMPBuilder.IMFTrackFileMetadata(headerPartitionBytes,
+                                hash,
+                                CompositionPlaylistBuilder_2016.defaultHashAlgorithm,
+                                mediaFile.getFile().getName(),
+                                resourceByteRangeProvider.getResourceSize())
+                );
+
+                if (copyTrackFiles) {
+                    File outputMediaFile = new File(outputDirectory.getAbsolutePath() + File.separator + mediaFile.getFile().getName());
+                    logger.info("Copying media file from\n{} to\n{}", mediaFile.getFile().getAbsolutePath(), outputMediaFile.getAbsolutePath());
+                    Files.copy(mediaFile.getFile().toPath(), outputMediaFile.toPath(), REPLACE_EXISTING);
+                }
+
+                // add to resources
+                logger.info("Adding file to resources: {}..", mediaFile.getFile().getName());
+                resources.add(
+                        new IMFTrackFileResourceType(
+                            UUIDHelper.fromUUID(UUID.randomUUID()),
+                            UUIDHelper.fromUUID(trackFileId),
+                            Arrays.asList(mediaFile.getSampleRate().getNumerator(), mediaFile.getSampleRate().getDenominator()),    // defaults to 1/1
+                            mediaFile.getIntrinsicDuration(),
+                            mediaFile.getEntryPoint(), // defaults to 0 if null
+                            mediaFile.getDuration(), // defaults to intrinsic duration if null
+                            mediaFile.getRepeatCount(), // defaults to 1 if null
+                            UUIDHelper.fromUUID(UUID.randomUUID()),   // used as the essence descriptor id
```

Comment:
There is a (minor) risk here, of simply using a random UUID for SourceEncoding, without doing additional record keeping.

Within a CPL, there should be a one-to-one mapping between `SourceEncoding` (specific Essence Descriptors) and `TrackFileId` (individual essence file).
But in a CPL, _multiple_ `Resources` can reference the same `TrackFile`. In these cases, when an essence file is reused, the different Resource entries would be expected to have the _same_ `SourceEncoding` value. By assigning this as a random UUID per `Resource`, this will create duplicate Essence Descriptors in the CPL.

This is actually a relatively common bug, and most IMF parsers _can_ handle it okay. But I believe the correct way to manage this would be to create some sort of dictionary, mapping essence files (`TrackFileId`) to essence descriptors (`SourceEncoding`).
Here is an [example where this is being done elsewhere in Photon](https://github.com/Netflix/photon/blob/master/src/main/java/com/netflix/imflibrary/writerTools/CompositionPlaylistBuilder_2016.java#L149-L171).

Reply:
Good catch! I'll add a lookup check. Looking at the CPL generated in the unit test where the same TrackFile is used twice, only one essence descriptor is created per track file, (and the Resource entries that point to the same TrackFile contain the same SourceEncoding) as hoped.